Create programming vignette with intro and tidy selection sections (#873)
tidyverse/tidyr: vignettes/in-packages.Rmd

### tidyselect interface

To provide an interface that works like the tidyr function that you're wrapping, you should pass the argument along using `{{ arg }}`. `{{ }}` is a special tidy eval operator that captures the expression supplied by the user and forwards it to another tidy eval-enabled function.
(Alternatively, you may want to use `any_of()` if you want the function to silently omit variables that are not present in the input data.)

```{r}
nest_egg <- function(df, cols) {
  nest(df, egg = {{ cols }})
}

nest_egg(mini_iris, -Species)
```

For more complicated functions, you might want to use tidyselect directly:

```{r}
sel_vars <- function(df, cols) {
  tidyselect::vars_select(names(df), {{ cols }})
}

sel_vars(mini_iris, -Species)
```

(Many tidyr functions use `...` so you can easily select many variables, e.g. `fill(df, x, y, z)`. I now believe that the disadvantages of this approach outweigh the benefits, and that this interface would have been better as `fill(df, c(x, y, z))`. For new functions that select columns, please just use a single argument and not `...`.)

> If you decide to support tidyselect syntax, I recommend re-exporting the select helpers so that users can use (e.g.) `starts_with()` without any extra work. There's no helper to [do this yet](https://github.com/r-lib/usethis/issues/868), but you can copy-and-paste the code that [tidyr uses](https://github.com/tidyverse/tidyr/blob/00a66b44d2f304a73e953dd73139db61ed9cc289/R/tidyr.R#L29-L53).

Reviewer: Is this re-exporting advice no longer necessary?

Reviewer 2: tidyselect 1.0.0 automatically makes them available in the right scope.

> Hopefully you've already adopted continuous integration for your package, in which `R CMD check` (which includes your own tests) is run on a regular basis, e.g. every time you push changes to your package's source on GitHub or similar. The tidyverse team currently relies most heavily on GitHub actions, so that will be our example. `usethis::use_github_action()` can help you get started.

Reviewer: ```suggestion
Hopefully you've already adopted continuous integration for your package, in which `R CMD check` (which includes your own tests) is run on a regular basis, e.g. every time you push changes to your package's source on GitHub or similar. The tidyverse team currently relies most heavily on GitHub Actions, so that will be our example. `usethis::use_github_action()` can help you get started.
```

> We recommend adding a workflow that targets the devel version of tidyr:

Reviewer: ```suggestion
We recommend adding a job that targets the devel version of tidyr:
```
We should either say "job" or stick with "build to your matrix", since workflow has a very specific meaning (= 1 Github Actions yaml file). 

Or, wait: are we actually recommending a totally separate workflow for this? Because now that I look at the yaml below, that's what it looks like.

Reviewer 2: Yeah, that's what I was trying to imply. I think it's better to keep your `R-cmd-check.yml` as pure as possible to make it easier to update in the future.

>   tidyr-devel workflow during the period preceding a major tidyr release that has 

Reviewer: ```suggestion
  tidyr-devel job to your build matrix during the period preceding a major tidyr release that has 
```
